website/docs/api/admin/feature-toggles-api-v2.md

```diff
@@ -599,3 +635,130 @@ http PATCH http://localhost:4242/api/admin/projects/default/features/demo/varian
+
+There is no way to update a user's role directly at the moment. Instead, if you want to change a user's role, first remove the user from the project and then add them back with the desired role.
+
+### Add a user to a project
+
+``` http
+POST /api/admin/projects/:projectId/users/:userId/roles/:roleId
+```
+
+This will add a user to a project and give the user a specified role within that project.
+
+#### URL parameters
+
+| Parameter   | Type    | Description                                                           | Example value     |
+|-------------|---------|-----------------------------------------------------------------------|-------------------|
+| `userId`    | integer | The ID of the user you want to add to the project.                    | `1`               |
+| `projectId` | string  | The id of the project to add the user to.                             | `"MyCoolProject"` |
+| `roleId`    | integer | The id of the role you want to assign to the new user in the project. | `7`               |
+
+
+#### Responses
+
+<details>
+<summary>Responses data</summary>
+
+##### 200 OK
+
+The user was added to the project with the specified role.
+
+``` json
+{ }
+```
+
+##### 400 Bad Request
+
+The user already exists in the project and cannot be added again:
+
+``` json
+[
+  {
+    "msg": "User already has access to project=<projectId>"
+  }
+]
+```
+
+</details>
+
+#### Example query
+
+The following query would add the user with ID 42 to the _MyCoolProject_ project and give them the role with ID 13.
+
+```bash
+http POST \
+http://localhost:4242/api/admin/projects/MyCoolProject/users/42/roles/13 \
+Authorization:$KEY
+```
+
+### Remove a user from a project
+
+``` http
+DELETE /api/admin/projects/:projectId/users/:userId/roles/:roleId
+```
+
+This will remove the specified from user from the project. The user _must_ have the role indicated by the `:roleId` URL parameter for the request to succeed.
+
+#### URL parameters
+
+| Parameter   | Type    | Description                                                           | Example value     |
+|-------------|---------|-----------------------------------------------------------------------|-------------------|
+| `userId`    | integer | The ID of the user you want to remove from the project.               | `1`               |
+| `projectId` | string  | The id of the project to remove the user from.                        | `"MyCoolProject"` |
+| `roleId`    | integer | The current role the of the user you want to remove from the project. | `7`               |
+
+#### Responses
+
+<details>
+<summary>Responses data</summary>
+
+:::caution
+If you provide the wrong role id for the user, but there is another user in the same project with the role you provided, you'll get a 200 OK response indicating success, but the user will not be removed.
+:::
```

Comment:
Is this correct? If so, it's really confusing. Any chance we can sneak in the changes necessary to make this comment _un_necessary?

Reply:
This goes back to the fact that DELETE ops should be idempotent. So if you DELETE role x from user y which happens and gives 200, it should give 200 if you try again as well, even though the user does not have this role anymore.

Reply:
Yeah, that's fair. I think the bigger issue here is that the endpoint acts differently depending on whether a different user in the same project has this role or not. It's been a while since I looked at the code, but I think @sighphyre and I discovered that, given users A and B and roles R and P:
- If user A and B both have role R and you delete R from A, you'll get a 200 OK
- If user A has role R, B has role P, and you delete R from A, you'll get a 200 OK
- If user A has role P, B has role R, and you delete R from A, you'll get a 200 OK
- If users A and B both have role P (and there is no other user in the project with role R), and you delete R from A, you'll get a 400 Bad Request error (or a different error).

The fact that this last case acts differently from all the others is surprising, and that's what I meant.

Reply:
@chriswk What's your take on this? If I understand correctly, the call will end up going via the `findProjectRole` function in `project-service.ts`, where it'll throw a NotFoundError if no user in thed project has that role: 

https://github.com/Unleash/unleash/blob/017bd75c9cd0967ce7678700bf094dd122b40ff5/src/lib/services/project-service.ts#L334-L338

Anyway, I'll update the note here before merging.

Reply:
Youssef and I have fixed the sequencing in the endpoint so that it does not check for the new role name before it has actually moved the user.

Reply:
Which means that in order to move the user now, you don't need an already existing user with the new role.

Reply:
Got it! So if I understand you correctly, that means that the aberrant response mentioned above has now been corrected? 😄